modules/misc/src/materials/ADDensity.C

```diff
@@ -26,10 +26,14 @@ ADDensity::validParams()
 ADDensity::ADDensity(const InputParameters & parameters)
   : ADMaterial(parameters),
     _coord_system(getBlockCoordSystem()),
-    _disp_r(adCoupledValue("displacements", 0)),
+    _disp_r(coupledComponents("displacements") ? adCoupledValue("displacements", 0) : _ad_zero),
     _initial_density(getParam<Real>("density")),
     _density(declareADProperty<Real>("density"))
 {
+  if (getParam<bool>("use_displaced_mesh"))
+    mooseError("ADDensity needs to act on an undisplaced mesh. Use of a displaced mesh leads to "
```

Comment:
please make this 

```
   paramError("use_displaced_mesh", "ADDensity needs to act on an undisplaced mesh. Use of a displaced mesh leads to incorrect gradient values");
```

Reply:
You're kind of still circling...if displacements aren't provided, and `use_displaced_mesh=true`, you'll still get an undisplaced behavior. It seems to me:
- `use_displaced_mesh` defaults true
- if `use_displaced_mesh=true`, require displacements
- if u`se_displaced_mesh=false`, require displacements not to be coupled

Reply:
would explicitly capture the behaviors you want with very clear input.

Reply:
No. `use_displaced_mesh` must **never** be true! The key is to look if a displaced mesh _exists_!

Reply:
```
if (_fe_problem.getDisplacedProblem() && coupledComponents("displacements") == 0)
  paramError("displacements", "Displacement variables need to be specified if a displaced mesh exists.);
```